### PR TITLE
LF: Change engine output from GenTransaction to VersionedTransaction

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
The versioning of the transaction is done in the engine. 

This is a prerequisite to implement #5164 

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/EngineTest.scala [hunk 1]
```diff
@@ -1609,6 +1609,10 @@ object EngineTest {
   private implicit def toName(s: String): Name =
     Name.assertFromString(s)
 
+  private implicit def toGenTransaction(
```
Comment (stefanobaghino-da) [inline daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/EngineTest.scala:1612]: I'd rather have this called explicitly. What's the advantage of having this as an implicit conversion? The explicit call looks very brief and understandable.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala [hunk 2]
```diff
@@ -51,9 +51,36 @@ final case class VersionedTransaction[Nid, +Cid](
...
+
+  def foldWithPathState[A, B](globalState0: A, pathState0: B)(
+      op: (A, B, Nid, GenNode.WithTxValue[Nid, Cid]) => (A, B)): A =
+    transaction.foldWithPathState[A, B](globalState0, pathState0)(op)
+
+  def localContracts[Cid2 >: Cid]: Map[Cid2, Nid] =
+    transaction.localContracts
+
+  def inputContracts[Cid2 >: Cid]: Set[Cid2] =
+    transaction.inputContracts
```
Comment (stefanobaghino-da) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala:75]: `transaction` is already a public member. If you want to keep it that way, I would rather keep those methods where they are and not redefining them here.

Comment (remyhaemmerle-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala:81]: You are right. 
The problem is that I do not want code external to LF relies on `GenTransaction` anymore.
This code should not try to build a transaction, but relies only on the Engine or the protobuf Decoder to produce transaction.

For testing,  one should use the `TransactionBuilder` I refactor in #6294

Comment (S11001001) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala:81]: This looks getter-lensy, so suggestion in branch `remy+s11-control-lf-versions`, specifically 66528f2e4ba0469961f4fb9098de7025d1bda974.

https://github.com/digital-asset/daml/blob/66528f2e4ba0469961f4fb9098de7025d1bda974/daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala#L288-L291

Comment (remyhaemmerle-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala]: :+1:

**Comment by remyhaemmerle-da** [comment]
cc @simonmaxen-da

**Pull request merged**